### PR TITLE
[Inductor] Add a device agnostic DeviceGuard class to inductor

### PR DESCRIPTION
**Pull request by brothergomez**
Summary: Currently although only in one place in inductor, the `device` context manager from the device interface is used . This PR creates an inductor specific `DeviceGuard` class for use in these cases, which keeps a reference to the `DeviceInterface` class which is defined and added out of tree. This then offloads the device specific work to the device interface, instead of having to define this logic on the device class which isn't strictly necessary for inductor.

Ideally I would have used the existing `DeviceGuard` class, but these are defined per device and don't work well with inductor's device agnostic/ out of tree compatible design. With the existing classes in mind, I am happy to take suggestions on the renaming of this class. 

Whilst I was there, I also took the opportunity to rename `gpu_device` to `device_interface` to clarify this is not necessarily a GPU. 

Test Plan: None currently, happy to add some. 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by colesbury** [comment]
@aakhundov, would you please review this?

File: torch/_dynamo/device_interface.py [hunk 1]
```diff
@@ -110,6 +118,31 @@ def get_compute_capability(device: _device_t = None):
...
+        self.idx = index
+        self.prev_idx = -1
+
+    def __enter__(self):
+        if self.idx is not None:
+            self.prev_idx = self.device_interface.exchange_device(self.idx)
+
+    def __exit__(self, type: Any, value: Any, traceback: Any):
+        if self.idx is not None:
+            self.idx = self.device_interface.maybe_exchange_device(self.prev_idx)
```
Comment (aakhundov) [inline torch/_dynamo/device_interface.py:142]: What is the purpose of setting `self.idx` here?

File: torch/_inductor/triton_heuristics.py [hunk 2]
```diff
@@ -356,9 +356,9 @@ def _precompile_config(self, cfg: Config, warm_cache_only_with_cc: Optional[int]
             )
 
         # load binary to the correct device
-        with self.gpu_device.device(compile_meta["device"]):  # type: ignore[attr-defined]
+        with DeviceGuard(self.device_interface, compile_meta["device"]):  # type: ignore[attr-defined]
             # need to initialize context
-            self.gpu_device.synchronize(self.gpu_device.current_device())
+            self.device_interface.synchronize(self.device.current_device())
```
Comment (aakhundov) [inline torch/_inductor/triton_heuristics.py:361]:
```suggestion
            self.device_interface.synchronize(self.device_interface.current_device())
```

File: torch/_inductor/triton_heuristics.py [hunk 3]
```diff
@@ -165,7 +165,7 @@ def __init__(
         self.device_type = (
             triton_meta["device_type"] if "device_type" in triton_meta else "cuda"
         )
-        self.gpu_device = get_interface_for_device(self.device_type)
+        self.device_interface = get_interface_for_device(self.device_type)
```
Comment (aakhundov) [inline torch/_inductor/triton_heuristics.py:168]: Maybe use `self.device` as a generalization of `self.gpu_device`?

**Review by aakhundov — COMMENTED** [review]
@brothergomez please add a unit test.

**Comment by brothergomez** [comment]
> @brothergomez please add a unit test.

Yep will do

File: torch/_inductor/triton_heuristics.py (hunk 3 above)
Comment (brothergomez) [reply torch/_inductor/triton_heuristics.py:168]: I feel that `device_interface` is clearer as this is what it's referring to? The actual `device` is a seperate thing.

**Comment by brothergomez** [comment]
I have added a test. Unsure whether to put it in inductor or dynamo subfolder. Not sure if the CUDA tests are necessary as well.

**Comment by brothergomez** [comment]
@aakhundov 
> Wondering, what is the purpose of setting self.idx here?

I think this question got lost or deleted? This` DeviceGuard` is a replica of the `torch.cuda._DeviceGuard` so kept it from there. Not sure how it is used there.

**Review by aakhundov — APPROVED** [review]

**Review by aakhundov — CHANGES_REQUESTED** [review]
LGTM. Let's also wait for the internal CI.

**Review by aakhundov — APPROVED** [review]

**Comment by brothergomez** [comment]
Hi, can this be merged @aakhundov ?

**Comment by aakhundov** [comment]
@brothergomez I think so: both OSS and internal CI look fine.

**Pull request closed without merging**